src/main/java/net/hydromatic/morel/foreign/CalciteForeignValue.java

```diff
@@ -85,24 +85,45 @@ private static String convert(boolean lower, String name) {
   public Object value() {
     final ImmutableList.Builder<List<Object>> fieldValues =
         ImmutableList.builder();
+
+    pushSchemaTablesIntoFieldValues(schema, fieldValues);
+
+    // Recursively walk sub-schemas and add their tables to fieldValues
+    SchemaPlus currentSchema = schema;
+    while (true) {
+      final Set<String> subSchemaNames = requireNonNull(currentSchema).getSubSchemaNames();
+      if (subSchemaNames.isEmpty()) {
+        break;
+      }
+      for (String subSchemaName : subSchemaNames) {
+        final SchemaPlus subSchema = requireNonNull(currentSchema.getSubSchema(subSchemaName));
+        System.out.println("CalciteForeignValue - processing sub-schema: " + subSchemaName);
+        // TODO: How to create a new ML record type for each sub-schema?
```

Comment:
You need to create a record type, and an instance of it. 

`TypeSystem.recordType(SortedMap)` creates the type. One field per sub-schema (or table) in the schema.

A record value is just a Java List. The entries correspond to the fields (which are always sorted by name). You should wrap each `Schema` object as a `CalciteForeignValue`, and each `Table` object as a `RelList`. (`CalciteForeignValue.value()` seems to do a lot of this already.)

Reply:
If you want an example of code creating an instance of a record, see how `Codes.TupleCode.eval(EvalEnv)` calls `Arrays.asList`. Tuples and records are the same at runtime (the only difference is that tuples have field names "1", "2", "3" etc whereas records have alphabetical names) so this piece of code is used for both.

Reply:
Ah okay great, thank you! Will try to implement this based on your info here 👍

Reply:
Hmm, I think I'm close but something here is breaking it:

```java
java.lang.ClassCastException: class java.lang.Integer cannot be cast to class java.util.List (java.lang.Integer and java.util.List are in module java.base of loader 'bootstrap')
```

```java
    public Type type(TypeSystem typeSystem) {
        return type(schema, typeSystem);
    }

    // New function (recursive)
    public Type type(SchemaPlus schema, TypeSystem typeSystem) {
        final ImmutableSortedMap.Builder<String, Type> fields =
                ImmutableSortedMap.orderedBy(RecordType.ORDERING);

        schema.getTableNames().forEach(tableName ->
                fields.put(convert(tableName),
                        toType(schema.getTable(tableName), typeSystem)));

        schema.getSubSchemaNames().forEach(subSchemaName -> {
            final SchemaPlus subSchema = schema.getSubSchema(subSchemaName);
            final ImmutableSortedMap.Builder<String, Type> subFields =
                    ImmutableSortedMap.orderedBy(RecordType.ORDERING);

            subSchema.getTableNames().forEach(tableName ->
                    subFields.put(convert(tableName),
                            toType(subSchema.getTable(tableName), typeSystem)));

            subSchema.getSubSchemaNames().forEach(subSubSchemaName -> {
              subFields.put(convert(subSubSchemaName),
                      type(subSchema.getSubSchema(subSubSchemaName), typeSystem));
            });

            fields.put(subSchemaName,
                    typeSystem.recordType(subFields.build()));
        });

        return typeSystem.recordType(fields.build());
    }
```

Reply:
I got it!
```
fieldMap = {todos={completed:bool, name:string} list}
fieldMap = {todos={completed:bool, name:string} list, users={age:int, name:string} list}
Out: val it = <relation> : {completed:bool, name:string} list
```